### PR TITLE
[4.0] [GSB] Improve handling of concrete types, type aliases, recursion

### PR DESCRIPTION
**Pull request by DougGregor**
**Explanation**: Improves our handling of concrete conformances, type aliases in protocols, and same-type-to-concrete constraints in the generic signature builder to make them all more consistent and robust.
**Scope**: Affects many uses of generic code; will reject a number of ill-formed cases that would previously crash the compiler.
**Radar**: SR-4295 / rdar://problem/31372308, SR-4757 / rdar://problem/31912838, SR-4786 / rdar://problem/31955862, SR-5014 / rdar://problem/32402482, SR-4737 / rdar://problem/31905232, plus a few other dupes.
**Risk**: Low-ish. The generic signature builder is key to all generics handling in Swift, so changes there can cause breakage... but it's also fairly well-tested via the standard library, so it's likely we'd have caught regressions.
**Testing**: Compiler regression testing, plus lots of new tests from the various fixed radars.

### CODE REVIEW

**Comment by DougGregor** [comment]
@swift-ci please test

**Comment by DougGregor** [comment]
@swift-ci please test compatibility

**Comment by DougGregor** [comment]
@swift-ci please test compatibility suite

**Comment by DougGregor** [comment]
@swift-ci please test source compatibility

**Comment by DougGregor** [comment]
The source-compatibility suite died with an exception:

FATAL: command execution failed
java.io.EOFException
	at java.io.ObjectInputStream$PeekInputStream.readFully(ObjectInputStream.java:2335)
	at java.io.ObjectInputStream$BlockDataInputStream.readShort(ObjectInputStream.java:2804)
	at java.io.ObjectInputStream.readStreamHeader(ObjectInputStream.java:802)
	at java.io.ObjectInputStream.<init>(ObjectInputStream.java:299)
	at hudson.remoting.ObjectInputStreamEx.<init>(ObjectInputStreamEx.java:48)
	at hudson.remoting.AbstractSynchronousByteArrayCommandTransport.read(AbstractSynchronousByteArrayCommandTransport.java:34)
	at hudson.remoting.SynchronousCommandTransport$ReaderThread.run(SynchronousCommandTransport.java:59)
Caused: java.io.IOException: Unexpected termination of the channel
	at hudson.remoting.SynchronousCommandTransport$ReaderThread.run(SynchronousCommandTransport.java:73)
Caused: java.io.IOException: Backing channel 'macOS-37' is disconnected.
	at hudson.remoting.RemoteInvocationHandler.channelOrFail(RemoteInvocationHandler.java:192)
	at hudson.remoting.RemoteInvocationHandler.invoke(RemoteInvocationHandler.java:257)
	at com.sun.proxy.$Proxy91.isAlive(Unknown Source)
	at hudson.Launcher$RemoteLauncher$ProcImpl.isAlive(Launcher.java:1043)
	at hudson.Launcher$RemoteLauncher$ProcImpl.join(Launcher.java:1035)
	at hudson.tasks.CommandInterpreter.join(CommandInterpreter.java:155)
	at hudson.tasks.CommandInterpreter.perform(CommandInterpreter.java:109)
	at hudson.tasks.CommandInterpreter.perform(CommandInterpreter.java:66)
	at hudson.tasks.BuildStepMonitor$1.perform(BuildStepMonitor.java:20)
	at hudson.model.AbstractBuild$AbstractBuildExecution.perform(AbstractBuild.java:779)
	at hudson.model.Build$BuildExecution.build(Build.java:206)
	at hudson.model.Build$BuildExecution.doRun(Build.java:163)
	at hudson.model.AbstractBuild$AbstractBuildExecution.run(AbstractBuild.java:534)
	at hudson.model.Run.execute(Run.java:1728)
	at hudson.model.FreeStyleBuild.run(FreeStyleBuild.java:43)
	at hudson.model.ResourceController.execute(ResourceController.java:98)
	at hudson.model.Executor.run(Executor.java:405)
Build step 'Execute shell' marked build as failure
ERROR: Step ‘Archive the artifacts’ failed: no workspace for swift-PR-source-compat-suite #247

**Comment by DougGregor** [comment]
@swift-ci please test source compatibility

**Comment by DougGregor** [comment]
@swift-ci please test source compatibility

**Comment by DougGregor** [comment]
@huonw , you reviewed the first 6 of these commits on master

File: include/swift/AST/GenericSignatureBuilder.h [hunk 1]
```diff
@@ -826,7 +835,7 @@ class GenericSignatureBuilder::RequirementSource final
     TypeBase *type;
 
     /// A protocol conformance used to satisfy the requirement.
-    ProtocolConformance *conformance;
+    void *conformance;
```
Comment (slavapestov) [inline include/swift/AST/GenericSignatureBuilder.h]: Why can't this just be a ProtocolConfomanceRef? You're including the header above

Comment (DougGregor) [inline include/swift/AST/GenericSignatureBuilder.h]: *sigh*. Non-POD types are really annoying to work with in unions.

File: validation-test/compiler_crashers_2_fixed/0042-rdar21775089.swift [hunk 2]
```diff
@@ -7,7 +7,7 @@ protocol MySequenceType {}
 protocol MyIndexableType {}
 
 protocol MyCollectionType : MySequenceType, MyIndexableType {
-  typealias SubSequence = MySlice<Self>
+  associatedtype SubSequence = MySlice<Self>
```
Comment (slavapestov) [inline validation-test/compiler_crashers_2_fixed/0042-rdar21775089.swift]: Why did you change the crasher test case?

**Comment by slavapestov** [comment]
LGTM!

File: validation-test/compiler_crashers_2_fixed/0042-rdar21775089.swift (hunk 2 above)
Comment (DougGregor) [inline validation-test/compiler_crashers_2_fixed/0042-rdar21775089.swift]: The test case itself was ill-formed because of clashing `typealias` definition, but we didn't catch it before because the GSB was ignoring the inconsistency. I believe the intent here was to use `associatedtype` with defaults, because the test itself was intended to be well-formed and go all the way to IRGen. This change makes the test case compile again.

**Comment by DougGregor** [comment]
@swift-ci please test

**Comment by DougGregor** [comment]
@swift-ci please test
